### DISCUSSION TITLE
Problems in running Pavlovian Conditioning code.

### DISCUSSION THREAD

**Discussion by JeongInHoi** [genn-team/genn#599] [GeNN questions] [answered]
I’m In-hoi Jeong and I’m MS-Ph.D. student of department of physics in Korea University. 
I'm trying to do some simulation about pavlovian conditioning using genn simulator.
I slightly modified your Pavlovian Conditioning Model code and try to run it. 
I want to set connectivity of the network manually so I changed

“e_e_pop = model.add_synapse_population("EE", "SPARSE_INDIVIDUALG", e_delay,
                                           "E", "E",
                                           izhikevich_stdp_model, stdp_params, stdp_init, {}, {},
                                           "DeltaCurr", {}, {},
                                          genn_model.init_connectivity("FixedProbability", fixed_prob_params))”
into 
“e_e_pop = model.add_synapse_population("EE", "SPARSE_INDIVIDUALG", e_delay,
                                           "E", "E",
                                           izhikevich_stdp_model, stdp_params, stdp_init, {}, {},
                                           "DeltaCurr", {}, {},
                                         None)
e_e_pop.set_sparse_connections(edge_e_e[0],edge_e_e[1])
”
In here, edge_e_e is a numpy array of connectivity which was saved from previous simulation.
I also double checked about matrix type and variable type in model definition.

Unfortunately, there is an “RuntimeError: synapseUpdate.cc: 335: cuda error 700: an illegal memory access was encountered”. 
I try to configure what happened and find out the error came from 

“CHECK_CUDA_ERRORS(cudaMemcpyToSymbolAsync(d_mergedPresynapticUpdateGroup0, &group, sizeof(MergedPresynapticUpdateGroup0), idx * sizeof(MergedPresynapticUpdateGroup0))”

It seems like that the cudaMemcpyToSymbolAsync function try to copy data from memory to gpu but failed because of error code 700. I guess error code 700 means there is a problem with variable type. 
It would be very helpful if you could provide me some advice about this failure.

Thank you very much for your time.

**Comment by tnowotny** [comment]
Hi, I think you might have broken it by passing ```None``` as the connectivity initialiser. The last argument of the connectivity initialiser is optional, so just do 
```
e_e_pop = model.add_synapse_population("EE", "SPARSE_INDIVIDUALG", e_delay,
"E", "E", izhikevich_stdp_model, stdp_params, stdp_init, {}, {},"DeltaCurr", {}, {})
``` 

(no "None")

**Reply by JeongInHoi** [reply]
Thank you for the quick answer.
I tried as what you said but the result is the same.(same error code appears)
And according to the GeNN documentation, connectivity initialiser is given as None by default. so I think there is some other problem.

**Comment by tnowotny** [comment]
Ah ... sorry ... would have been too easy, @neworderofjamie will have a deeper look.

**Reply by neworderofjamie** [reply]
@JeongInHoi, you are totally right that ``None`` has the same effect.  Could you share the numpy file you're loading ``edge_e_e`` from?

**Reply by JeongInHoi** [reply]
Here is the link for the numpy file.
https://github.com/JeongInHoi/data-for-question-genn.git
I used "edge_e_e = np.load('ind_e_e_120.npy')" for load the edge_e_e

**Reply by neworderofjamie** [reply]
Thank you! What ``size_scale_factor`` are you passing to the ``get_params`` function?

**Reply by JeongInHoi** [reply]
size_scale_factor is 1 and weight_scale_factor is 1.0
I used 1600 excitatory neuron and 400 inhibitory neuron for the simulation.

**Reply by JeongInHoi** [reply]
I also upload simulation code for the case you need it.
https://github.com/JeongInHoi/data-for-question-genn.git

**Reply by neworderofjamie** [reply, marked as answer]
So I think your issue is that when you save your EI connectivity at https://github.com/JeongInHoi/data-for-question-genn/blob/main/izhikevich_pavlovian_PB_consolidation0616.py#L240 you add ``no_exc`` to the postsynaptic indices but, when you load it at https://github.com/JeongInHoi/data-for-question-genn/blob/main/common_PB_consolidation0616.py#L305-L308 you don't subtract this off again, meaning that the connectivity points to non-existent neurons, causing the error you found. 

Also, your simulations are downloading the connectivity and synaptic weights and writing them to disk _extremely_ frequently. This is very inefficient, especially for the connectivity as it does not change after the model is loaded so only needs downloading and writing to disk once!

-

**Reply by JeongInHoi** [reply]
Thanks a lot! Now it works and i also erase the code about downloading connectivity part:)
